ppl/lake/walk.go

```diff
@@ -111,32 +114,40 @@ func tsDirEntriesToChunks(ctx context.Context, lk *Lake, filterSpan nano.Span, t
 			m[id] = s
 		}
 	}
-	var chunks []chunk.Chunk
+
+	var mu sync.Mutex
+	chunks := make([]chunk.Chunk, 0, len(m))
+	group, ctx := errgroup.WithContext(ctx)
+
 	for id, seen := range m {
 		if !seen.meta || !seen.data {
 			continue
 		}
-		dir := tsDir.path(lk)
-		mdPath := chunk.MetadataPath(dir, id)
-		b, err := lk.immfiles.ReadFile(ctx, mdPath)
-		if err != nil {
-			return nil, err
-		}
+		id := id
+		group.Go(func() error {
+			dir := tsDir.path(lk)
+			mdPath := chunk.MetadataPath(dir, id)
+			b, err := lk.immfiles.ReadFile(ctx, mdPath)
+			if err != nil {
+				if zqe.IsNotFound(err) {
+					return nil
+				}
+				return fmt.Errorf("failed to read chunk metadata from %v: %w", mdPath, err)
+			}
 
-		md, err := chunk.UnmarshalMetadata(b, lk.DataOrder)
-		if err != nil {
-			if zqe.IsNotFound(err) {
-				continue
+			md, err := chunk.UnmarshalMetadata(b, lk.DataOrder)
+			if err != nil {
+				return err
 			}
-			return nil, fmt.Errorf("failed to read chunk metadata from %v: %w", mdPath, err)
-		}
-		chunk := md.Chunk(dir, id)
-		if !filterSpan.Overlaps(chunk.Span()) {
-			continue
-		}
-		chunks = append(chunks, chunk)
+
+			mu.Lock()
+			chunks = append(chunks, md.Chunk(dir, id))
+			mu.Unlock()
+			return nil
+		})
 	}
-	return chunks, nil
+
+	return chunks, group.Wait()
```

Comment:
This seems like a race since `chunks` will be evaluated before `group.Wait()`.
```suggestion
	err := group.Wait()
	return chunks, err
```

Reply:
Good point! Considering it is possible for errgroup to return Wait before all go routines finish execution (in the case of a go routine returning an error, or the context getting cancelled), it's probably best to do this:
```suggestion
        if err := group.Wait(); err != nil {
            return nil, err
        }
        return chunks, nil
```

Reply:
Either approach seems OK to me.

In general, I think it's fine to return a non-nil value along with a non-nil error under the assumption that callers will ignore the value unless the error is nil or the documentation otherwise (as with io.Reader.Read).

> Considering it is possible for errgroup to return Wait before all go routines finish execution (in the case of a go routine returning an error, or the context getting cancelled),

This isn't quite right. `group.Wait` won't return until all functions passed to `group.Go` have returned.

Reply:
Yep you are right. In that case i'll write it as you suggested.